# __ECC102 &ndash; Fall 2019-2020 &ndash; Homework 04__

## Student Name and ID

Please enter your name and your student ID below programmatically!

In [3]:
student_name = ""
student_id   = ""

## SUBMISSION INSTRUCTIONS

You will submit your own copy of this Jupyter Lab notebook according to the instructions below:

- You will upload the completed version of this notebook to the DropBox site that will be provided to you along with the Facebook announcement.

- See the upload site for the __deadline__. 

- __Late homeworks will NOT be accepted!__ 

- Name your notebook file using your first name, middle name--if you have one, your
  last name, and your 8-digit student ID, separated by individual underscore
  characters. All letters MUST be in LOWERCASE.

  DO NOT USE any PUNCTUATION characters in your filename. USE ONLY letters a-z
  and digits 0-9, except for the underscore characters to join different parts
  of your filename (first/middle/last name, student ID).

  For example, if your name is `John Milton Smith` and your student ID is
  `20173847`, then your file should be named 

                          john_milton_smith_20173847.ipynb

  You may abbreviate your middle name and hence your file may also be named

                          john_m_smith_20173847.ipynb

- DO NOT include any other information in the name of your file that you will upload!

- <font color="blue">__READ ALL INSTRUCTIONS CAREFULLY BEFORE STARTING TO SOLVE EACH PROBLEM__</font>

## HOMEWORK UPLOAD

Upload your homework here:
    
> https://www.dropbox.com/request/tScgBBFHDtToWfWO8j4o

## Questions
---

There are 11 questions on this homework. Best of luck!

----
QUESTION 1

----

Write a Python function `sum_positive_even_numbers_1( number_list )` that takes a `list` of `int` values, `number_list`, and returns the sum of only the _even_ values in `number_list`. 

- Use __looping__ to implement your solution.

For example,

    sum_positive_even_numbers_1( [1, 2, 3, 4, 5, 6, 7, 8] )
    
should return `20`.

In [12]:
def sum_positive_even_numbers_1(number_list):
    sum = 0
    for n in number_list:
        if n % 2 == 0:
            sum = n + sum
           
    return sum
print(sum_positive_even_numbers_1( [1, 2, 3, 4, 5, 6, 7, 8] ))
    
    


20


----
QUESTION 2

----

Similar to Question 1, write a Python function `sum_positive_even_numbers_2( number_list )` that takes a `list` of `int` values, `number_list`, and returns the sum of only the _even_ values in `number_list`. 

- Use __list comprehension__ to implement your solution.

For example,

    sum_positive_even_numbers_2( [1, 2, 3, 4, 5, 6, 7, 8] )
    
should return `20`.

In [1]:
def sum_positive_even_numbers_2(number_list):
    sum = 0
    n = 0
    number_list = [n for n in number_list if n % 2 == 0]
    [(n + sum) for (n, sum) in zip(number_list[:-1], number_list[1:])]
    return sum + n
print(sum_positive_even_numbers_2( [1, 2, 3, 4, 5, 6, 7, 8] ))


0


----
QUESTION 3

----

Write a function called `asc_range( start, end, increment=1 )` that returns a range of integer values as a list given the `start` value, `end` value, and the (positive) `increment`, where $\mathtt{start} \le \mathtt{end}$. Your function should work similar to the way the `range(.)` function works.

- __You are NOT allowed to you any range-producing ready-made function!__ 

- You will use a `while`-loop to create your own value range. The default value of `increment` will be `1`. That is, if not specified, `increment` will automatically have the value `1`.

For example,
        
    asc_range( 5, 12 )

should return
       
    [5, 6, 7, 8, 9, 10, 11]
    
and
    
    asc_range( 5, 12, 3 ) # Note the increment argument
    
should return

    [5, 8, 11]

In [11]:

def asc_range(start,end,increment = 1):
    i = start  
    while i < end:
        yield i
        i += increment
        
print list(asc_range(5,12))
print list(asc_range(5,12,3))    
             


[5, 6, 7, 8, 9, 10, 11]
[5, 8, 11]


----
QUESTION 4

----

Resolve Question 3, but this time use a `for`-loop instead.

In [8]:

ar = []
def asc_range(start,end):
    for i in [start]:
        if i == end:
           
            ar.append(end)
            
            return ar
        else:
            if i != 7:
                ar.append(i)
                
            return asc_range(i + 1, end)
        
print(list(asc_range(5,11)))

       

[5, 6, 8, 9, 10, 11]


----
QUESTION 5

----

Write a function called `celsius_to_fahrenheit( degrees )`, which converts the `degrees` value from _degrees_ to _Fahrenheit_. This function should take an `int` value in Celcius and return the corresponding temperature in Fahrenheit.

- Using a `for`-loop, demonstrate that your function works as specified for 10 different input values.

In [1]:
def getConvertTo():
    which = input("Enter Selection: ")
    while which != "F" and which != "C":
        which = input("Enter Selection ")
    return which
which = getConvertTo()

def FahrenToCelsius(start,end):
    print("Fahrenheit", "Celsius")
    
    for temp in range (start,end, +1):
        converted_temp = (temp - 32) * 5/9
        print(" ",format(temp)," ",format(int(converted_temp)))
        
temp_start = int(input("Enter temprature to convert: "))
temp_end = int(input("Enter ending temprature to convert: "))

FahrenToCelsius(temp_start, temp_end)




Enter Selection:  F


NameError: name 'F' is not defined

----
QUESTION 6

----

Write a function called `fahrenheit_to_celsius( fahrenheit )`, which converts the `fahrenheit` value from _fahrenheit_ to _celsius_. This function should take an `int` value in Fahrenheit and return the corresponding temperature in Celcius.

- Using a `for`-loop, demonstrate that your function works as specified for 10 different input values.

In [ ]:
def getConvertTo():
    which = input("Enter Selection: ")
    while which != "F" and which != "C":
        which = input("Enter Selection ")
    return which
which = getConvertTo()

def FahrenToCelsius(start,end):
    print("Fahrenheit", "Celsius")
    
    for temp in range (start,end, +1):
        converted_temp = (temp - 32) * 5/9
        print(" ",format(temp)," ",format(int(converted_temp)))
        
temp_start = int(input("Enter temprature to convert: "))
temp_end = int(input("Enter ending temprature to convert: "))

FahrenToCelsius(temp_start, temp_end)

----
QUESTION 7

----

Write a Python function called `uppercase_last_letters( string )` that takes a string and _uppercases_ the last letter of each word in that string and returns the resulting string.

For example,

    uppercase_last_letters( "This is a test String" )
    
should return
    
    "ThiS iS A tesT StrinG"
    
- Do not break up the string into words by splitting
- You should detect the end of a string with the existence of a space character instead

In [10]:
# Write your answer here
def capitalize(string):
    string, result = string.title(), ""
    for word in string.split():
        result += word[:-1] + word[-1].upper() + " "
    return result[:-1]     

print capitalize("This is a test string")




ThiS IS A TesT StrinG


----
QUESTION 8

----

Write a Python _function_ named `count_strings( items )`, which takes a list of values and returns how many _strings_ that list contains at the top level. For example, for

    items = ['apple', 55, 1.2, 'banana', lambda a: a, 
             'pear', None, 'cherry', """Hello world!""", -2,
             '''The Who''', ("a", 5), [("a", "5"), ("b", 3)]]
the call

    count_strings( items )
    
should return `6`.
    
Note that you will have to use the `type(.)` call to determine the type of a given value or object. For example,

    >>> type( 1.2 )
    float

In [68]:
# Write your answer here
def count_strings(items):
    i = 0
    for item in items: 
        if(type(item) == str):
            i += 1
    return i

t =  ['apple', 55, 1.2, 'banana', lambda a: a, 'pear', None, 'cherry', """Hello world!""", -2, '''The Who''', ("a", 5), [("a", "5"), ("b", 3)]]

print(count_strings(t))



    

6


----
QUESTION 9

----

Do the same as in ___Question 9___, but this time use a `for-`loop, instead of a `while-`loop.

In [11]:
# Write your answer here
def count_strings(items):
    i = 0
    count = 0 
    while i < len(items):
        if(isinstance(items[i],str)):
            count += 1
        i += 1
            
    return count   
            
                  
t =  ['apple', 55, 1.2, 'banana', lambda a: a, 'pear', None, 'cherry', """Hello world!""", -2, '''The Who''', ("a", 5), [("a", "5"), ("b", 3)]]

print(count_strings(t))





6


In [ ]:
----
QUESTION 10

----

Write a Python function called `std_deviation( values )` that takes a `list` of numeric values (`int` and/or `float`) and return the _standard deviation_, $s$, of these values.

$${\displaystyle s = {\sqrt {\frac {\sum _{i=1}^{N}(x_{i}-{\overline {x}})^{2}}{N - 1}}} }$$

where $x_1$, $x_2$, $x_3$, $\cdots$ $x_n$ are observed values in the sample data&mdash;that will be provided in the `values` parameter to the above function,
$\scriptstyle {\displaystyle\overline {x}}$ is the _mean_ (or average) value of the these observed values and
$N$ is the number of sample observations.

- You will not use any ready-made function that computes the standard deviation. You will write this function from scratch

For example,

    sample = [1, 2, 3, 4, 5] 
    std_deviation( sample )
    
should return

    1.5811388300841898

You can double-check your work using the `numpy` or `statistics` modules.

In [7]:
import numpy as np

sample = [1, 2, 3, 4, 5] 
np.std( sample, ddof=1 ) # Note the ddof=1 must be passed!

1.5811388300841898

In [8]:
import statistics

sample = [1, 2, 3, 4, 5] 
statistics.stdev( sample )

1.5811388300841898

In [67]:
# Write your answer here
import statistics 

def std_deviation( values ):
  
    values = [1, 2, 3, 4, 5] 
    return statistics.stdev(values)
values = [1, 2, 3, 4, 5] 
print("Standard Deviation of values is % s " % (statistics.stdev(values)))




ImportError: No module named statistics

----
QUESTION 11

----

Write a Python function called `correlation_coefficient( sample_a, sample_b )` which takes two `list`s of sample numeric values of the same length and returns the correlation coefficient (a `float`) for these two samples.

- See this page for the algorithm for computing the correlation coefficient, $r_x$:

   > https://www.mathsisfun.com/data/correlation.html
    
    This page also provides a detailed example.
    
You can check your work using the `numpy` module's `corrcoef(.)` function as follows:

In [10]:
sample_a = [14.2, 16.4, 11.9, 15.2, 18.5, 22.1, 19.4, 25.1, 23.4, 18.1, 22.6, 17.2]
sample_b = [ 215, 325, 185, 332, 406, 522, 412, 614, 544, 421, 445, 408]

np.corrcoef( sample_a, sample_b )

array([[1.        , 0.95750662],
       [0.95750662, 1.        ]])

You can see from the returned value above that the $r_x$ value is `0.95750662`, which matches the value given in the above webpage.

In [3]:
# Write your answer here
import numpy as np
def correlation_coefficient(sample_a , sample_b):
    
    return np.corrcoef(sample_a,sample_b)

sample_a = [14.2, 16.4, 11.9, 15.2, 18.5, 22.1, 19.4, 25.1, 23.4, 18.1, 22.6, 17.2]
sample_b = [ 215, 325, 185, 332, 406, 522, 412, 614, 544, 421, 445, 408]   

print(np.corrcoef( sample_a, sample_b))

    
    
    
    
    
    


[[1.         0.95750662]
 [0.95750662 1.        ]]


---

<center>END OF HOMEWORK</center>